### DISCUSSION TITLE
Error using OpenCL API

### DISCUSSION THREAD

**Discussion by muzapparjan** [dotnet/Silk.NET#383] [General] [answered]
Here is my code:

```
cl = CL.GetApi();
uint platformCount;
cl.GetPlatformIDs(0, null, &platformCount);
```

And it threw exception: System.DivideByZeroException:“Attempted to divide by zero.” while executing GetPlatformIDs(xxx).

Then I tried to write as the script in Silk.Net/src/Lab/CLMultiplication/Program.cs:

```
cl = CL.GetApi();
IntPtr platform;
cl.GetPlatformIDs(1, &platform, null);
```

And it threw the same exception.

Environment: 
* Windows 10 64bit
* Using Silk.Net.OpenGL & ..OpenCL & ..Windowing & ..Common & ..Input & ..OCL Extention KHR

**Comment by Perksey** [comment]
Can you please give the complete exception including the stack trace? I can't reproduce here.

**Comment by muzapparjan** [comment]
System.DivideByZeroException
  HResult=0x80020012
  Message=Attempted to divide by zero.
  Source=<无法计算异常源>
  StackTrace:
<无法计算异常堆栈跟踪>

![image](https://user-images.githubusercontent.com/24959643/103308808-5215d880-4a4e-11eb-81b8-5201501e3255.png)
![image](https://user-images.githubusercontent.com/24959643/103308953-b0db5200-4a4e-11eb-9aa1-aeb72f50cf45.png)

**Comment by Perksey** [comment]
Can you right click on the Visual Studio Call stack pane and enable the "Show external code" option?

**Comment by Perksey** [comment]
Oh nevermind I can see what I need to see.

**Comment by Perksey** [comment]
Can you try updating to Silk.NET 2.0 Preview 5?

**Comment by muzapparjan** [comment]
Emmmm
I am a green hand...
Can you please tell me how to update?

**Comment by muzapparjan** [comment]
I downloaded it directly on NuGet

**Comment by Perksey** [comment]
- Right click on your project, and select `Manage NuGet packages`
- Head over to the updates tab, and select `Include prerelease`

Now the updated verisons should pop up. Alternatively if that's giving you issues:
- open the .csproj file in a text editor
- replace every reference of "1.9.0" with "2.0.0-preview5"

**Comment by muzapparjan** [comment]
Wow really thanks!!!

**Comment by muzapparjan** [comment]
...
It still throws the same exception

**Comment by Perksey** [comment]
Can you send me the stack trace again? The $exception.StackTrace string is what clued me in last time. It's impossible that that's the same, if it is you didn't upgrade correctly.

**Comment by muzapparjan** [comment]
at Silk.NET.OpenCL.CL.GetPlatformIDs(UInt32 num_entries, IntPtr* platforms, UInt32* num_platforms)
at Study.ImageProcessing.Test.Tests.CLTrialTask.OnLoad() in E:\\CSharp\\Study\\Study.ImageProcessing.Test\\Tests\\CLTrialTask.cs:line 52
at Silk.NET.Windowing.Internals.ViewImplementationBase.Initialize()
at Silk.NET.Windowing.WindowExtensions.Run(IView view)
at Study.ImageProcessing.Test.Tests.CLTrialTask.Run(String[] args, Object[] inputs) in E:\\CSharp\\Study\\Study.ImageProcessing.Test\\Tests\\CLTrialTask.cs:line 31
at Study.ImageProcessing.Test.Entrance.Main(String[] args) in E:\\CSharp\\Study\\Study.ImageProcessing.Test\\Entrance.cs:line 9"

**Comment by muzapparjan** [comment]
![image](https://user-images.githubusercontent.com/24959643/103309840-1b8d8d00-4a51-11eb-86f8-19b14debb0b9.png)

**Comment by muzapparjan** [comment]
My repository named "Study" is the problem project

**Comment by Perksey** [comment]
Your repo does not have the images required.

**Comment by muzapparjan** [comment]
Just put a picture named input.jpg at the exe's location or retarget the path in CLTrialTask.cs line 46

**Comment by muzapparjan** [comment]
Nothing special here(*^_^*)

**Comment by Perksey** [comment]
![image](https://user-images.githubusercontent.com/11160611/103311089-4f48c280-4a11-11eb-9424-24b2d44c0de7.png)

**Comment by muzapparjan** [comment]
Wow didn't it throw exceptions at calling clGetPlatformIDs API?

**Comment by Perksey** [comment]
Nope.

**Comment by muzapparjan** [comment]
I cried

**Comment by muzapparjan** [comment]
![image](https://user-images.githubusercontent.com/24959643/103311272-d79c8700-4a54-11eb-9cb9-326080583731.png)

**Comment by Perksey** [comment]
It's strange because we don't do any division. Can you try right clicking on the call stack window and enabling "Show external code"?

**Comment by muzapparjan** [comment]
![image](https://user-images.githubusercontent.com/24959643/103311392-382bc400-4a55-11eb-95e4-52cf80b43e8d.png)

**Comment by muzapparjan** [comment]
未知 means unknown

**Comment by Perksey** [comment]
What's the text at the very top of the call stack?

![image](https://user-images.githubusercontent.com/11160611/103311632-c894e500-4a12-11eb-91fe-e3dab1a0f815.png)

**Comment by muzapparjan** [comment]
调用堆栈：Call Stack
名称：Name
托管到本机的转换：Transition from managed to host(???)

**Comment by Perksey** [comment]
This looks to be an OpenCL bug, converting to discussion.

Can you try updating your drivers?

**Comment by Perksey** [comment]
Also, can you try **Enable native code debugging** and see if anything changes?
![image](https://user-images.githubusercontent.com/11160611/103311857-57a1fd00-4a13-11eb-905f-70727e3aebe8.png)

**Comment by muzapparjan** [comment]
I enabled all native code debugging options in solution and the result is:
![image](https://user-images.githubusercontent.com/24959643/103312359-bf7a3700-4a57-11eb-9f0f-e7f70c5b83fd.png)

**Comment by muzapparjan** [comment]
帧状态：Frame State
非用户代码，已加载符号：Isn't written by user, symbol is loaded
无可用源：No Available Source

**Comment by muzapparjan** [comment]
I updated nvidia driver from v457.xx to v460.89 but it still not work

**Reply by Perksey** [reply]
That's a shame. This unfortunately is a driver issue, so there's no real way I can diagnose this further.

We did some experimentation and the runtime is smart enough to recognise the error in the unmanaged code as a DivideByZeroException because of a specific interrupt it generates (`#DE`). For example, this throws the right exception:
```
struct div_t
{
    private int a;
    private int rem;
    public override string ToString()
    {
        return $"{a} rem {rem}";
    }
}
using var lib = new UnmanagedLibrary("MSVCR100.DLL");
Console.WriteLine(((delegate* unmanaged[Stdcall]<int, int, div_t>)lib.LoadFunction("div"))(100, 0));
```

So the problem is definitely with the driver. There's not much more help I can provide I'm afraid.

**Comment by muzapparjan** [comment, marked as answer]
It doesn't matter. Thank you for your time
